### ISSUE TITLE
The 	preview changes dialog edit field is not readable within the NuGet client using Access technology

### ISSUE THREAD

**Issue by SeanFarrow** [NuGet/Home#12837]
### NuGet Product Used

Visual Studio Package Management UI

### Product Version

6.x

### Worked before?

no

### Impact

It's more difficult to complete my work

### Repro Steps & Context

An issue exists where the edit field in the Preview Changes dialog that indicates what actions will be taken by NuGet within a project does not read correctly with both NVDA and JAWS on Windows.

I would expect this to be navigable like any other standard edit control.

The reason this isn't working is that this appears to be a custom control of some form.

In order to replicate this, ensure you are running either NVDA or JAWS. Then try updating a package within any project. The actions NuGet will take are visible on the screen but are not readable by access technologies.

If someone could point me to where this is in the source code, I'll gladly have a look and see how easy the fix might be.

### Verbose Logs

```shell
No logs needed.
```

**Comment by kartheekp-ms** [comment]
> An issue exists where the edit field in the Preview Changes dialog that indicates what actions will be taken by NuGet within a project does not read correctly with both NVDA and JAWS on Windows.

Can you please provide a screenshot so that I can understand which edit field in the preview changes dialog are you referring to? The one that I am aware of is preview changes dialog while applying analyzer code fixes. From NuGet tooling, the only preview changes window, I am aware of is the dialog box that is displayed asking users for permission when installing a package.

**Comment by SeanFarrow** [comment]
Hi,

It’s the preview dialog that displays what packages will be installed, updated and removed.

I’ve found the problem and am working through what the potential fixes might be.

Thanks,
Sean.

**Comment by donnie-msft** [comment]
> I’ve found the problem and am working through what the potential fixes might be.

Thank you for taking the initiative! If you don't mind sharing what you're looking at, I might be able to help and/or save some unnecessary work for you. At this point, I don't understand the problem and I don't know which VS version you're experiencing it in.

**Comment by SeanFarrow** [comment]
Good morning Donnie,

This is with Visual studio 2022 version 17.7.2 and NuGet package manager version 6.7.0. I have noticed this hasn’t worked for a while though.

Essentially, if you look at the following file:
https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PreviewWindow.xaml

You will see the following:
      <ScrollViewer
        HorizontalScrollBarVisibility="Auto"
        IsTabStop="True"
        AutomationProperties.LabeledBy="{Binding ElementName=_changesText, Mode=OneWay}">
        <ItemsControl
          x:Name="_changesItems"
          ItemsSource="{Binding Path=PreviewResults}"
          IsTabStop="False">
          <ItemsControl.ItemTemplate>
            <DataTemplate>
              <StackPanel
                Margin="6, 6">
                <TextBlock
                  x:Name="_changeName"
                  AutomationProperties.Name="{Binding Text}"
                  FontSize="{Binding ElementName=_self,Path=FontSize,Converter={StaticResource Font122PercentSizeConverter}}"
                  Text="{Binding Path=Name}" />

                
                <StackPanel
                  Margin="0,8"
                  Visibility="{Binding Path=Deleted,Converter={StaticResource EnumerableToVisibilityConverter}}">
                  <TextBlock
                    x:Name="_uninstalledPackages"
                    AutomationProperties.Name="{Binding Text}"
                    FontWeight="Bold"
                    Text="{x:Static nuget:Resources.Label_UninstalledPackages}" />
                  <ItemsControl
                    Margin="10,0,0,0"
                    ItemsSource="{Binding Path=Deleted}"
                    IsTabStop="False">
                    <ItemsControl.ItemTemplate>
                      <DataTemplate>
                        <TextBlock Text="{Binding}" AutomationProperties.Name="{Binding AutomationName}" />
                      </DataTemplate>
                    </ItemsControl.ItemTemplate>
                  </ItemsControl>
                </StackPanel>

                
                <StackPanel
                  Margin="0,8"
                  Visibility="{Binding Path=Updated,Converter={StaticResource EnumerableToVisibilityConverter}}">
                  <TextBlock
                    x:Name="_updatedPackages"
                    AutomationProperties.Name="{Binding Text}"
                    FontWeight="Bold"
                    Text="{x:Static nuget:Resources.Label_UpdatedPackages}" />
                  <ItemsControl
                    Margin="10,0,0,0"
                    ItemsSource="{Binding Path=Updated}"
                    IsTabStop="False">
                    <ItemsControl.ItemTemplate>
                      <DataTemplate>
                        <TextBlock Text="{Binding}" AutomationProperties.Name="{Binding AutomationName}" />
                      </DataTemplate>
                    </ItemsControl.ItemTemplate>
                  </ItemsControl>
                </StackPanel>

                
                <StackPanel
                  Margin="0,8"
                  Visibility="{Binding Path=Added,Converter={StaticResource EnumerableToVisibilityConverter}}">
                  <TextBlock
                    x:Name="_installedPackages"
                    AutomationProperties.Name="{Binding Text}"
                    FontWeight="Bold"
                    Text="{x:Static nuget:Resources.Label_InstalledPackages}" />
                  <ItemsControl
                    Margin="10,0,0,0"
                    ItemsSource="{Binding Path=Added}"
                    IsTabStop="False">
                    <ItemsControl.ItemTemplate>
                      <DataTemplate>
                        <TextBlock Text="{Binding}" AutomationProperties.Name="{Binding AutomationName}" />
                      </DataTemplate>
                    </ItemsControl.ItemTemplate>
                  </ItemsControl>
                </StackPanel>
              </StackPanel>
            </DataTemplate>
          </ItemsControl.ItemTemplate>
        </ItemsControl>
      </ScrollViewer>

The issue here is that access technology is unable to pick up the fact that there are multiple text blocks containing information as they are all inside StackPanels, do not have tab stops and also do not seem to have the correct ui automation properties set up.

Even if the correct UI automation properties were set up, I doubt this would give a nice experience, as the text is not navigable.
Admittedly, we can copy the text to the clipboard, but I would respectfully suggest that this isn’t optimal as it does slow you down—you then have to paste the text somewhere before you can read it.

Could an alternative layout be considered, such as just a read-only edit field? This is only my initial suggestion and doesn’t have to be the final outcome.

I hope that helps.

Kind regards
Sean.

**Comment by rperez030** [comment]
@SeanFarrow, I completely agree with your assessment of the issue. This is really a usability issue, not an accessibility issue. The content of that dialog can be read with either JAWS's scan cursor or NVDA's object navigator, as well as with Narrrator's scan mode. Arguably it would be much easier to have that content presented in a read only edit field or a listbox containing tasks where one can tab to, and that's why I consider it as a usability issue.

**Comment by SeanFarrow** [comment]
@rperez030,

I completely agre.

The other two options you have are a HTML-like browser window or a read-only tree view, with nodes for updated, added and removed packages.

Out of all of these options, the treeview seems the most sensible. I am discounting the browser window as it is really diffecult to get back to the rest of the dialog as access technology capture the focus to an extend.

What is the best way to move this forward?

Thanks,
Sean.

**Comment by rperez030** [comment]
If you use JAWS, virtualizing the current window with Insert alt +W would pretty much produce the same result as having the content rendered in a webview, for example. There is an [NVDA add-on ](https://github.com/ruifontes/virtualReview
) That has very similar functionality.

**Comment by jeffkl** [comment]
Team Triage: @donnie-msft can give a cost estimate to fix this so we can decide if immediate action is needed?
